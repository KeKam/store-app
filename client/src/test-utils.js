import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
