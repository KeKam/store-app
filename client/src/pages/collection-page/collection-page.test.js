import React from 'react';
import '@testing-library/jest-dom';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import { render, screen } from '../../test-utils';
import {
  startFetchCollection,
  fetchCollectionSuccess,
} from '../../redux/collection/collection.actions';
import CollectionPage from './collection-page';

const history = createMemoryHistory();

describe('CollectionPage component', () => {
  const mockMatch = {
    isExact: true,
    params: {
      categoryId: 'spring',
    },
    path: '/collection',
    url: '/collection',
  };

  const mockState = {
    categories: {
      spring: {
        id: 1,
        items: [
          {
            id: 1,
            imageUrl: 'www.testImage.com',
            name: 'Tulips',
            price: 20,
          },
        ],
        routeName: 'spring',
        title: 'Spring',
      },
    },
    isFetching: false,
  };

  it('should render CollectionPage component', () => {
    const { asFragment } = render(
      <Router history={history}>
        <CollectionPage match={mockMatch} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByTestId('container')).toBeInTheDocument();
  });

  it('should render CollectionCategoryPage component', () => {
    history.push('/collection/spring');

    render(
      <Router history={history}>
        <CollectionPage match={mockMatch} />
      </Router>,
      { initialState: { collection: mockState } }
    );

    expect(screen.getByRole('heading', { name: 'SPRING' })).toBeInTheDocument();
    expect(screen.getByTestId('image')).toBeInTheDocument();
    expect(screen.getByTestId('title')).toHaveTextContent('Tulips');
    expect(screen.getByTestId('price')).toHaveTextContent('20 €');
    expect(screen.getByTestId('link')).toHaveTextContent('Sign in required');
  });

  it('should dispatch startFetchCollection action on render', () => {
    const { getDispatchedActions } = render(
      <Router history={history}>
        <CollectionPage match={mockMatch} />
      </Router>
    );

    expect(getDispatchedActions()).toContainEqual(startFetchCollection());
  });

  it('should render CollectionOverview component', () => {
    const collection = {
      spring: {
        id: 1,
        items: [
          {
            id: 1,
            imageUrl: 'www.testImage.com',
            name: 'Tulips',
            price: 20,
          },
        ],
        routeName: 'spring',
        title: 'Spring',
      },
    };

    history.push('/collection');

    const { dispatch } = render(
      <Router history={history}>
        <CollectionPage match={mockMatch} />
      </Router>,
      { initialState: { collection: mockState } }
    );

    dispatch(fetchCollectionSuccess(collection));

    expect(screen.getByRole('heading', { name: 'SPRING' })).toBeInTheDocument();
    expect(screen.getByTestId('image')).toBeInTheDocument();
    expect(screen.getByTestId('title')).toHaveTextContent('Tulips');
    expect(screen.getByTestId('price')).toHaveTextContent('20 €');
    expect(screen.getByTestId('link')).toHaveTextContent('Sign in required');
  });
});
