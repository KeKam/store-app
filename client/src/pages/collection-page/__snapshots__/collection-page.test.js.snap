// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionPage component should render CollectionPage component 1`] = `
<DocumentFragment>
  <div
    class="sc-cSHVUG eDNIQr"
    data-testid="container"
  />
</DocumentFragment>
`;
